js/ui/autocomplete.jsx

```diff
@@ -0,0 +1,209 @@
+
+const defaultStyles = StyleSheet.create({
+  container: {
+    width: '100%',
+    alignItems: 'center',
+  },
+  input: {
+    height: 40,
+    width: '90%',
+    margin: 12,
+    borderWidth: 1,
+    padding: 10,
+    borderRadius: 10,
+    borderColor: '#ccc',
+  },
+  resultList: {
+    width: '90%',
+  },
+  resultItem: {
+    padding: 10,
+    borderBottomWidth: 1,
+    borderBottomColor: '#ccc',
+  },
+  selectButton: {
+    borderColor: '#000275',
+    borderWidth: 1,
+    paddingHorizontal: 14,
+    paddingVertical: 7,
+    borderRadius: 10,
+    justifyContent: 'center',
+  },
+  selectButtonText: {
+    color: '#000275',
+    fontSize: 12,
+  },
+  addressText: {
+    fontSize: 14,
+    color: '#000',
+  },
+  footerContainer: {
+    flexDirection: 'row',
+    alignItems: 'center',
+    padding: 10,
+    alignSelf: 'flex-start',
+  },
+  footerText: {
+    marginTop: 2,
+    marginRight: 8,
+  },
+  logo: {
+    width: 60,
+    height: 18,
+    resizeMode: 'contain',
+  },
+});
+
+const defaultAutocompleteOptions = {
+  debounceMS: 200, // Debounce time in milliseconds
+  threshold: 3, // Minimum number of characters to trigger autocomplete
+  limit: 8, // Maximum number of autocomplete results
+  placeholder: 'Search address', // Placeholder text for the input field
+  disabled: false,
+  showSelectButton: false,
+  buttonText: 'Select',
+};
+
+const autocompleteUI = ({ options = {}, onSelect, location, style = {} }) => {
+  const [query, setQuery] = useState('');
+  const [results, setResults] = useState([]);
+  const [isOpen, setIsOpen] = useState(false);
+  const config = { ...defaultAutocompleteOptions, ...options };
+
+  const styles = {
+    container: StyleSheet.compose(defaultStyles.container, style.container),
+    input: StyleSheet.compose(defaultStyles.input, style.input),
+    resultList: StyleSheet.compose(defaultStyles.resultList, style.resultList),
+    resultItem: StyleSheet.compose(defaultStyles.resultItem, style.resultItem),
+    footerContainer: StyleSheet.compose(defaultStyles.footerContainer, style.footerContainer),
+    footerText: StyleSheet.compose(defaultStyles.footerText, style.footerText),
+  };
+
+  const fetchResults = useCallback(async (searchQuery) => {
+    const params = {
+      query: searchQuery,
+      limit: config.limit,
+    };
+
+    if (location && location.latitude && location.longitude) {
+      params.near = location;
+    }
+
+    Radar.autocomplete(params).then((result: any) => {
+      setResults(result.addresses);
+      setIsOpen(true);
+    }).catch((error) => {
+      if (config.onError && typeof config.onError === 'function') {
+        config.onError(error);
+      }
+    });
+  }, [config]);
+
+  const handleInput = useCallback(
+    (text) => {
+      setQuery(text);
+      if (text.length < config.threshold) {
+        return;
+      }
+      const timer = setTimeout(() => {
+        fetchResults(text);
+      }, config.debounceMS);
+
+      return () => clearTimeout(timer);
```

Comment:
Does the timer get cleared if another input change comes in before the previous one is fired? It seems like this will still fire all the API calls, just after a delay (a new input change should interrupt any pending request, to reduce API call volume)

Reply:
Ah, indeed we need to cancel previous pending reqs. Here's the fix: https://github.com/radarlabs/react-native-radar/pull/232/commits/3e93f55d15ba2e1bf43609bdb99b94a4b2599fff